pymc3/smc/smc.py

```diff
@@ -1,279 +1,438 @@
         )
-    elif kernel.lower() == "metropolis":
-        likelihood_logp = logp_forw([model.datalogpt], variables, shared)
-
-    if parallel and cores > 1:
-        pool = mp.Pool(processes=cores)
-        priors = pool.starmap(prior_logp, [(sample,) for sample in posterior])
-        likelihoods = pool.starmap(likelihood_logp, [(sample,) for sample in posterior])
-    else:
-        priors = [prior_logp(sample) for sample in posterior]
-        likelihoods = [likelihood_logp(sample) for sample in posterior]
-
-    priors = np.array(priors).squeeze()
-    likelihoods = np.array(likelihoods).squeeze()
-
-    while beta < 1:
-        beta, old_beta, weights, sj = calc_beta(beta, likelihoods, threshold)
-
-        model.marginal_likelihood *= sj
-        # resample based on plausibility weights (selection)
-        resampling_indexes = np.random.choice(np.arange(draws), size=draws, p=weights)
-        posterior = posterior[resampling_indexes]
-        priors = priors[resampling_indexes]
-        likelihoods = likelihoods[resampling_indexes]
-
-        # compute proposal distribution based on weights
-        covariance = _calc_covariance(posterior, weights)
-        proposal = MultivariateNormalProposal(covariance)
-
-        # compute scaling (optional) and number of Markov chains steps (optional), based on the
-        # acceptance rate of the previous stage
-        if (tune_scaling or tune_steps) and stage > 0:
-            scaling, n_steps = _tune(
-                acc_rate, proposed, tune_scaling, tune_steps, scaling, max_steps, p_acc_rate
+        self.any_discrete = self.discrete.any()
+        self.all_discrete = self.discrete.all()
+
+    def initialize_population(self):
+        """
+        Create an initial population from the prior distribution
+        """
+        population = []
+        var_info = OrderedDict()
+        if self.start is None:
+            init_rnd = sample_prior_predictive(
+                self.draws, var_names=[v.name for v in self.model.unobserved_RVs], model=self.model
+            )
+        else:
+            init_rnd = self.start
+
+        init = self.model.test_point
+
+        for v in self.variables:
+            var_info[v.name] = (init[v.name].shape, init[v.name].size)
+
+        for i in range(self.draws):
+
+            point = Point({v.name: init_rnd[v.name][i] for v in self.variables}, model=self.model)
+            population.append(self.model.dict_to_array(point))
+
+        self.posterior = np.array(floatX(population))
+        self.var_info = var_info
+
+    def setup_kernel(self):
+        """
+        Set up the likelihood logp function based on the chosen kernel
+        """
+        shared = make_shared_replacements(self.variables, self.model)
+        self.prior_logp = logp_forw([self.model.varlogpt], self.variables, shared)
+
+        if self.kernel.lower() == "abc":
+            warnings.warn(EXPERIMENTAL_WARNING)
+            simulator = self.model.observed_RVs[0]
```

Comment:
Could you add an assertion here to make sure the length of observed_RVs is equal 1, otherwise throw an error?